### PR TITLE
lax.custom_linear_solve primitive

### PR DESCRIPTION
**Pull request by shoyer**
A follow-on to `lax.root` from https://github.com/google/jax/pull/1339. See the tests for a simple example of defining gradients for an iterative linear solver.

Eventually, I expect to use this for implementing iterative solvers (e.g., GMRES, CG) from `scipy.sparse.linalg`, which in turn we will use for the linear solves needed for gradients in `scipy.optimize.root`.

### CODE REVIEW

File: jax/lax/lax_control_flow.py [hunk 1]
```diff
@@ -977,4 +975,130 @@ def _root_batch(args, dims, **params):
...
+
+
+_Solves = collections.namedtuple('_Solves', 'forward tangent cotangent')
+
+def linear_solve(matvec, b, forward_solve, tangent_solve=None,
+                 cotangent_solve=None, symmetric=False):
+  """Differentiably solve the linear map matvec(x)=b for x.
+
+  Required invariant:
+      x = solve(matvec, b)
```
Comment (dougalm) [inline jax/lax/lax_control_flow.py]: `solve` should be `forward_solve`. Perhaps also describe what `(co)tangent_solve` is supposed to be and what invariants it observes?

File: jax/lax/lax_control_flow.py [hunk 2]
```diff
@@ -977,4 +975,130 @@ def _root_batch(args, dims, **params):
...
+
+  params, _ = split_list(primals, [num_consts])
+  params_dot, b_dot = split_list(tangents, [num_consts])
+
+  matvec = partial(core.jaxpr_as_fun(jaxpr), *params)
+  matvec_dot = partial(
+      _tangent_linear_map, core.jaxpr_as_fun(jaxpr), params, params_dot
+  )
+
+  if all(tangent is ad_util.zero for tangent in params_dot):
```
Comment (dougalm) [inline jax/lax/lax_control_flow.py]: What happens if `params_dot` or `b_dot` are a mixture of zero and not-zero? Seems like it might blow up.

File: jax/lax/lax_control_flow.py [hunk 3]
```diff
@@ -977,4 +975,130 @@ def _root_batch(args, dims, **params):
...
+def linear_solve(matvec, b, forward_solve, tangent_solve=None,
+                 cotangent_solve=None, symmetric=False):
+  """Differentiably solve the linear map matvec(x)=b for x.
+
+  Required invariant:
+      x = solve(matvec, b)
+      error = matvec(x) - b
+      assert all(error == 0)
+  """
+  if tangent_solve is None:
```
Comment (dougalm) [inline jax/lax/lax_control_flow.py]: Are these two cases exercised by the tests?

**Review by dougalm — APPROVED** [review]

File: jax/lax/lax_control_flow.py [hunk 4]
```diff
@@ -974,14 +977,141 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
...
+
+linear_solve_p = core.Primitive('linear_solve')
+linear_solve_p.multiple_results = True
+linear_solve_p.def_impl(_linear_solve_impl)
+linear_solve_p.def_abstract_eval(_linear_solve_abstract_eval)
+ad.primitive_jvps[linear_solve_p] = _linear_solve_jvp
+xla.initial_style_translations[linear_solve_p] = xla.lower_fun(
+    _linear_solve_impl, initial_style=True)
+ad.primitive_transposes[linear_solve_p] = _linear_solve_transpose_rule
+batching.deftraced(linear_solve_p)
```
Comment (jekbradbury) [inline jax/lax/lax_control_flow.py]: Do you have a sense of whether this will run into #1249?

Comment (shoyer) [reply jax/lax/lax_control_flow.py]: Yeah, I was wondering the same thing...

Comment (shoyer) [reply jax/lax/lax_control_flow.py]: Yeah, I was wondering the same thing...

Comment (shoyer) [reply jax/lax/lax_control_flow.py]: Indeed, this most definitely suffers from the same issue. It appears we will need to write custom batching rules for `root` and `linear_solve`. I think this would be pretty doable, but now I'm leaving this as a "TODO".

File: jax/lax/lax_control_flow.py (hunk 1 above)
Comment (shoyer) [reply jax/lax/lax_control_flow.py]: Done, I wrote a full docstring.

File: jax/lax/lax_control_flow.py (hunk 2 above)
Comment (shoyer) [reply jax/lax/lax_control_flow.py]: This should be handled now. I'm not entirely happy with the test coverage, but struggled to figure out how to come up with a test case that could involve a mixture of zeros/non-zeros -- I think I would need a three argument primitive?

File: jax/lax/lax_control_flow.py (hunk 3 above)
Comment (shoyer) [reply jax/lax/lax_control_flow.py]: It is now :)

**Comment by shoyer** [comment]
Please take another look. I rewrote everything so `solve`/`transpose_solve` are now evaluated to JAXprs before calling the primitive, so they work on closures. Also added a bunch more tests/docs/validation.

File: jax/lax/lax_control_flow.py [hunk 5]
```diff
@@ -974,14 +976,194 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
...
+      assert matvec(x) == b  # not checked
+
+  Args:
+    matvec: linear function to invert. Must be differentiable.
+    b: constant right handle side of the equation. May be any nested structure
+      of arrays.
+    solve: higher level function that solves for solution to the linear
+      equation, i.e., ``matvec(solve(matvec, x)) == x`` for all ``x`` of the
+      same form as ``b``. This function need not be differenatiable.
+    transpose_solve: higher level function for solving the transpose linear
```
Comment (gehring) [inline jax/lax/lax_control_flow.py]: Is `transpose_solve` used when no derivatives are ever computed, i.e., just used as a regular solver? I'm just thinking that it would be good for requirement for `transpose_solve` should be as generous as possible.

Though this could be overly cautious on my part. Is there a case where someone would use `linear_solve` and not need derivatives given that the user already has to provide a `solve` function?

File: jax/lax/lax_control_flow.py [hunk 6]
```diff
@@ -974,14 +976,194 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
...
+  if actual_shapes != expected_shapes:
+    actual_shape_tree = tree_unflatten(tree, actual_shapes)
+    act_shape_tree = tree_unflatten(tree, actual_shapes)
+    raise ValueError('{}() output shapes must match {}, got {} and {}'
+                     .format(func_name, expected_name,
+                             tree_unflatten(tree, actual_shapes),
+                             tree_unflatten(tree, expected_shapes)))
+
+
+def linear_solve(matvec, b, solve, transpose_solve=None, symmetric=False):
```
Comment (gehring) [inline jax/lax/lax_control_flow.py]: Does it make sense to call it `linear_solve` if the user has to provide a function to solve the linear system? The purpose of this function is clear to me, but I'm trying to put myself in a new user's shoes.

A new user might not be familiar with the subtleties of automatic differentiation. This is similar to our `lax.root` discussion. It might be good to make sure it is clear that the purpose is to define "appropriate" derivatives. I think that something with `def*` prefix, ala `defjvp`, makes it clear but the docs alone could possibly do the trick. I'm afraid that a new user would read `linear_solve` and just assume to know what it does. An "unusual" name (like something with a `def*` prefix) would avoid users assuming they know what the function does just from the name.
WDYT?

File: jax/lax/lax_control_flow.py (hunk 5 above)
Comment (shoyer) [reply jax/lax/lax_control_flow.py]: `transpose_solve` only gets used if a VJP needs to be calculated, e.g., for calculating backwards gradients with `jax.grad`. It's certainly possible to use this function for only the JVP rule, though I think this would be a little unusual.

I'll see if I can delay the error from not defining `transpose_solve` until the VJP rule without making things much more complex.

File: jax/lax/lax_control_flow.py (hunk 6 above)
Comment (shoyer) [reply jax/lax/lax_control_flow.py]: I agree that an unusual name would be a good idea here. I'm leaning towards `differentiable_linear_solve` per our discussion over in https://github.com/google/jax/issues/1448. (I'll also definitely add clearer docs explaining why this is useful!)

Comment (shoyer) [reply jax/lax/lax_control_flow.py]: Maybe `custom_linear_solve`?

File: jax/lax/lax_control_flow.py (hunk 5 above)
Comment (gehring) [reply jax/lax/lax_control_flow.py]: Okay, that's what I thought! I agree that it is not worth changing it if it would bring in a lot of additional complexity. This can always be looked at if this becomes an actual limitation for someone rather than just a theoretical issue. Though if it's an easy fix, might as well do it now!

File: jax/lax/lax_control_flow.py (hunk 6 above)
Comment (gehring) [reply jax/lax/lax_control_flow.py]: Our discussion is now a little split but I'm on board!

> I'll also definitely add clearer docs explaining why this is useful!

Great! Though, there is no need to over do it here, imo. 2-3 brief sentences should be enough for people to know what to look up if they don't understand. More detailed docs might deserve a standalone implicit diff page similar to async dispatch and co, that explains why backprop through all the things isn't the only solution.

File: jax/lax/lax_control_flow.py [hunk 7]
```diff
@@ -974,14 +976,201 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
...
+  This function allows for overriding or defining gradients for a linear
+  solve directly via implicit differentiation at the solution, rather than by
+  differenting *through* the solve operation. This can sometimes be much faster
+  or more numerically stable, or differentiating through the solve operation
+  may not even be implemented (e.g., if ``solve`` using ``lax.while_loop``).
+
+  Required invariant:
+      x = solve(matvec, b)  # solve the linear equation
+      assert matvec(x) == b  # not checked
+
```
Comment (gehring) [inline jax/lax/lax_control_flow.py]: > Maybe `custom_linear_solve`?

I like it! It's reminiscent of `jax.custom_gradient` and should cue users to pay attention to the docs. Also, the additional docs look great!

File: jax/lax/lax_control_flow.py (hunk 5 above)
Comment (shoyer) [reply jax/lax/lax_control_flow.py]: Done

**Comment by shoyer** [comment]
This is ready for a final (?) review.

File: jax/lax/lax_control_flow.py [hunk 8]
```diff
@@ -974,14 +976,209 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
...
+def _split_linear_solve_args(args, const_lengths):
+  params_list = split_list(args, list(const_lengths))
+  return _LinearSolveTuple(*params_list[:-1]), params_list[-1]
+
+
+def _transpose_function(linear_fun, primals):
+  """Transpose a linear function."""
+  # TODO(shoyer): can we use something more direct than the vjp machinery?
+  # It's particularly awkward that we need the second argument to give
+  # particular values of the primals, which are entirely arbitrary.
```
Comment (dougalm) [inline jax/lax/lax_control_flow.py]: Good idea. It might be a new transformation, but one that shares a lot of machinery with the existing jvp/transpose.

**Review by dougalm — APPROVED** [review]
